app/routes/goal_routes.py

```
@@ -1 +1,128 @@
+        
+    except KeyError as error:
+        response = {"details": "Invalid data"}
+        abort(make_response(response, 400))
+
+    db.session.add(new_goal)
+    db.session.commit()
+
+    return {"goal": new_goal.to_dict()}, 201
+
+@bp.get("")
+def get_all_goals():
+    query = db.select(Goal)
+
+    title_param = request.args.get("title")
+    if title_param:
+        query = query.where(Goal.title.ilike(f"%{title_param}%"))
+
+    # Handle sorting
+    sort_param = request.args.get("sort")
+    if sort_param == "asc":
+        query = query.order_by(Goal.title.asc())
+    elif sort_param == "desc":
+        query = query.order_by(Goal.title.desc())
+    else:
+        query = query.order_by(Goal.id)
+    
+    # Execute query
+    goals = db.session.scalars(query).all()
+    goals_response = []
+
+    for goal in goals:
+        goals_response.append({
+            "id": goal.id,
+            "title": goal.title
+        })
+
+    return goals_response, 200
+
+@bp.get("/<goal_id>")
+def get_one_goal(goal_id):
+    goal = validate_goal(goal_id)
+    return {"goal": goal.to_dict()}
+
+@bp.put("/<goal_id>")
+def update_goal(goal_id):
+    goal = validate_goal(goal_id)
+    request_body = request.get_json()
+
+    goal.title = request_body["title"]
+    db.session.commit()
+
+    return Response(status=204, mimetype="application/json")
+
+@bp.delete("/<goal_id>")
+def delete_goal(goal_id):
+    goal = validate_goal(goal_id)
+    db.session.delete(goal)
+    db.session.commit()
+
+    return Response(status=204, mimetype="application/json")
+
+@bp.post("/<goal_id>/tasks")
+def create_task_to_goal(goal_id):
+    goal = validate_goal(goal_id)
+
+    request_body = request.get_json()
+    task_ids = request_body.get("task_ids")
+
+    if not task_ids or not isinstance(task_ids, list):
+        return {"message": f"Invalid request: missing task_ids"}, 400
+
+    # Fetch tasks from DB
+    query = db.select(Task).where(Task.id.in_(task_ids))
```

Great use of `in_` to simplify validating the tasks! 

Currently if a task id is passed that does not exist, we silently ignore it. If we wanted to let a user know that an invalid task id was passed, what would we need to change or update?

---

to handle the case with invalid task_ids passed we could add a check under the request lines:
` # Fetch tasks from DB
    query = db.select(Task).where(Task.id.in_(task_ids))
    tasks = db.session.scalars(query).all()`
 This check would check which task IDs from the request do exist in the database. if any is missing the function could return 404 error. The code could look something like this:
`found_ids = [task.id for task in tasks]
 missing_ids = list(set(task_ids) - set(found_ids))

 if missing_ids:
        return {"message": f"Invalid task_ids: {missing_ids} not found"}, 404`